File: Cpp/Capacity_To_Ship_Packages_Within_D_Days.cpp
class Solution {
public:
    bool isPossible(std::vector<int>& weights, int D, int capacity) {
        int days_needed = 1;
        int current_weight = 0;

        for (int w : weights) {
            if (current_weight + w <= capacity) {
                current_weight += w;
            } else {
                days_needed++;
                current_weight = w;
            }
        }
        return days_needed <= D;
    }

    int shipWithinDays(std::vector<int>& weights, int D) {
        int low = 0;
        long long high = 0;

        for (int w : weights) {
            low = std::max(low, w);
            high += w;
        }

        int ans = high;
        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (isPossible(weights, D, mid)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }
};

File: Rust/Capacity_To_Ship_Packages_Within_D_Days.rs
impl Solution {
    fn is_possible(weights: &Vec<i32>, d: i32, capacity: i32) -> bool {
        let mut days_needed = 1;
        let mut current_weight = 0;

        for &weight in weights {
            if current_weight + weight <= capacity {
                current_weight += weight;
            } else {
                days_needed += 1;
                current_weight = weight;
            }
        }
        days_needed <= d
    }

    pub fn ship_within_days(weights: Vec<i32>, d: i32) -> i32 {
        let mut low = *weights.iter().max().unwrap_or(&0);
        let mut high: i32 = weights.iter().sum();
        let mut ans = high;
        while low <= high {
            let mid_capacity = low + (high - low) / 2;
            
            if Self::is_possible(&weights, d, mid_capacity) {
                ans = mid_capacity;
                high = mid_capacity - 1;
            } else {
                low = mid_capacity + 1;
            }
        }
        ans
    }
}